### PR TITLE
Avoid using DenseElementsAttr::getRawData

### PR DESCRIPTION
**Pull request by sdasgup3**
fixes https://github.com/openxla/stablehlo/issues/83

The PR implements  the underlying data storage of the tensor using an `mlir::HeapAsmResourceBlob` object represents a mutable blob of data. This provides two improvements:
1. No need for using `DenseElementsAttr::getRawData()` which cannot be relied  to provide contiguous representation of underlying data in cases like boolean and splats.
2.  An `mlir::HeapAsmResourceBlob` object, being mutable, simplifies how we store the underlying Tensor data.

### CODE REVIEW

File: docs/interpreter.md
Comment (burmako) [inline docs/interpreter.md:10]: Consider simplifying since the name of the class was just called out on the previous line. E.g. "The data storage contains a mutable blob of data...".

File: stablehlo/reference/Tensor.cpp
Comment (burmako) [inline stablehlo/reference/Tensor.cpp:68]: In the current design, `Tensor::Tensor(const Tensor &other)` and `Tensor &Tensor::operator=(const Tensor &other)` don't involve copying memory ([see details](https://github.com/llvm/llvm-project/blob/dfbaf90043841d18f5871e12b10c674b9b3abf99/llvm/include/llvm/ADT/IntrusiveRefCntPtr.h#L174-L196)), but it looks like in the proposed design, they both involve copying. Do we consider that to be a downside? If yes, is there something that can be done about it?

**Review by burmako — CHANGES_REQUESTED** [review]
Here are some thoughts on this PR. I'll leave commenting on using `HeapAsmResourceBlob` to Mehdi, since I haven't been following that very closely.

File: stablehlo/reference/Tensor.cpp
Comment (sdasgup3) [reply stablehlo/reference/Tensor.cpp:58]: Done.

File: stablehlo/reference/Ops.cpp
Comment (sdasgup3) [reply stablehlo/reference/Ops.cpp:32]: done

File: docs/interpreter.md
Comment (sdasgup3) [reply docs/interpreter.md]: done

File: stablehlo/reference/Tensor.cpp
Comment (sdasgup3) [reply stablehlo/reference/Tensor.cpp]: Thanks for catching this. This allows lots of unnecessary copies. Fixed it by reference counting the internal storage.

File: stablehlo/reference/Tensor.h
Comment (sdasgup3) [reply stablehlo/reference/Tensor.h]: Now these methods should move to the reference counted class `Buffer`.

File: stablehlo/reference/Ops.cpp
Comment (burmako) [inline stablehlo/reference/Ops.cpp:19]: We don't need these two includes then?

File: stablehlo/reference/Tensor.h
Comment (burmako) [inline stablehlo/reference/Tensor.h:50]: Any reason to change the name from `getData` to `getImmutableData`? Perhaps let's stick to whatever convention already exists upstream?

File: stablehlo/reference/Tensor.cpp
Comment (burmako) [inline stablehlo/reference/Tensor.cpp:338]: Let's leave a comment about why we are using `uint16_t` to store `f16` and `bf16`.

**Review by burmako — COMMENTED** [review]
Will leave the rest of the review to Mehdi, including commenting on memory management of blobs.

**Review by burmako — APPROVED** [review]
LGTM from my perspective, still needs a review from Mehdi.

File: docs/interpreter.md
Comment (joker-eph) [inline docs/interpreter.md:13]: Some ownership / lifetime aspect should be documented: a Tensor is an immutable object and copying a tensor doesn't copy the underlying buffer?
(also the Tensor class could be documented, I don't see the header being modified here)

**Review by joker-eph — APPROVED** [review]

File: stablehlo/reference/Tensor.cpp
Comment (joker-eph) [inline stablehlo/reference/Tensor.cpp:71]: If a Tensor is immutable, I'm not sure what do we do with this ctor?

**Comment by sdasgup3** [comment]
Let us port the discussion on mutability of tensor from https://github.com/openxla/stablehlo/pull/101 to this PR.

**Comment by sdasgup3** [comment]
We will be addressing the comments from @joker-eph via [ticket](https://github.com/openxla/stablehlo/issues/233). Planning to merge this as this will unblock writing tests with boolean types and splat.

**Pull request merged**
